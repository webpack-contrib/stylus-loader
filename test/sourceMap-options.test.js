import path from "path";
import fs from "fs";

import {
  compile,
  getCodeFromBundle,
  getCodeFromStylus,
  getCompiler,
  getErrors,
  getWarnings,
} from "./helpers";

jest.setTimeout(10000);

describe('"sourceMap" options', () => {
  it('should generate source maps when value is "true"', async () => {
    const testId = "./source-map.styl";
    const compiler = getCompiler(testId, {
      sourceMap: true,
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const { css, map } = codeFromBundle;

    map.sourceRoot = "";
    map.sources = map.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(map.sourceRoot, source))).toBe(true);

      return path
        .relative(path.resolve(__dirname, ".."), source)
        .replace(/\\/g, "/");
    });

    const codeFromStylus = await getCodeFromStylus(testId, {
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });

    expect(codeFromBundle.css).toBe(codeFromStylus.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps when the "devtool" value is "source-map"', async () => {
    const testId = "./source-map.styl";
    const compiler = getCompiler(
      testId,
      {
        stylusOptions: {
          paths: ["test/fixtures/paths"],
        },
      },
      {
        devtool: "source-map",
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const { css, map } = codeFromBundle;

    map.sourceRoot = "";
    map.sources = map.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(map.sourceRoot, source))).toBe(true);

      return path
        .relative(path.resolve(__dirname, ".."), source)
        .replace(/\\/g, "/");
    });

    const codeFromStylus = await getCodeFromStylus(testId, {
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });

    expect(codeFromBundle.css).toBe(codeFromStylus.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate source maps when value is "true" and the "devtool" value is "false"', async () => {
    const testId = "./source-map.styl";
    const compiler = getCompiler(
      testId,
      {
        sourceMap: true,
        stylusOptions: {
          paths: ["test/fixtures/paths"],
        },
      },
      {
        devtool: false,
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const { css, map } = codeFromBundle;

    map.sourceRoot = "";
    map.sources = map.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(map.sourceRoot, source))).toBe(true);

      return path
        .relative(path.resolve(__dirname, ".."), source)
        .replace(/\\/g, "/");
    });
    const codeFromStylus = await getCodeFromStylus(testId, {
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });

    expect(codeFromBundle.css).toBe(codeFromStylus.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps when value is "false"', async () => {
    const testId = "./source-map.styl";
    const compiler = getCompiler(testId, {
      sourceMap: false,
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const { css, map } = codeFromBundle;
    const codeFromStylus = await getCodeFromStylus(testId, {
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });

    expect(codeFromBundle.css).toBe(codeFromStylus.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps when the "devtool" value is "false"', async () => {
    const testId = "./source-map.styl";
    const compiler = getCompiler(
      testId,
      {
        stylusOptions: {
          paths: ["test/fixtures/paths"],
        },
      },
      {
        devtool: false,
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const { css, map } = codeFromBundle;
    const codeFromStylus = await getCodeFromStylus(testId, {
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });

    expect(codeFromBundle.css).toBe(codeFromStylus.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should not generate source maps when value is "false" and the "devtool" value is "source-map"', async () => {
    const testId = "./source-map.styl";
    const compiler = getCompiler(
      testId,
      {
        sourceMap: false,
        stylusOptions: {
          paths: ["test/fixtures/paths"],
        },
      },
      {
        devtool: "source-map",
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const { css, map } = codeFromBundle;
    const codeFromStylus = await getCodeFromStylus(testId, {
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });

    expect(codeFromBundle.css).toBe(codeFromStylus.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toBeUndefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should generate nested source maps when value is "true"', async () => {
    const testId = "./source-map/index.styl";
    const compiler = getCompiler(testId, {
      sourceMap: true,
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const { css, map } = codeFromBundle;

    map.sourceRoot = "";
    map.sources = map.sources.map((source) => {
      expect(path.isAbsolute(source)).toBe(true);
      expect(source).toBe(path.normalize(source));
      expect(fs.existsSync(path.resolve(map.sourceRoot, source))).toBe(true);

      return path
        .relative(path.resolve(__dirname, ".."), source)
        .replace(/\\/g, "/");
    });

    const codeFromStylus = await getCodeFromStylus(testId, {
      stylusOptions: {
        paths: ["test/fixtures/paths"],
      },
    });

    expect(codeFromBundle.css).toBe(codeFromStylus.css);
    expect(css).toMatchSnapshot("css");
    expect(map).toMatchSnapshot("source map");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("should work and allow to override source maps options", async () => {
    const testId = "./basic.styl";
    const stylusOptions = {
      sourcemap: {
        comment: true,
        inline: true,
      },
    };
    const compiler = getCompiler(testId, { stylusOptions });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle(stats, compiler);
    const codeFromStylus = await getCodeFromStylus(testId, { stylusOptions });

    expect(codeFromBundle.css).toBe(codeFromStylus.css);
    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});
