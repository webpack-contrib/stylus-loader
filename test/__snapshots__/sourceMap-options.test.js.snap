// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"sourceMap" options should generate nested source maps when value is "true": css 1`] = `
".other {
  font-family: serif;
}
body {
  font: 12px Helvetica, Arial, sans-serif;
}
a.button {
  -webkit-border-radius: 5px;
  -moz-border-radius: 5px;
  border-radius: 5px;
}
.some-class {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should generate nested source maps when value is "true": errors 1`] = `[]`;

exports[`"sourceMap" options should generate nested source maps when value is "true": source map 1`] = `
{
  "mappings": "AAAA;EACC,aAAY,MAAZ;;ACID;EACE,MAAmB,kCAAnB;;AAEF;EAPE,uBAAsB,IAAtB;EACA,oBAAmB,IAAnB;EACA,eAAc,IAAd;;ACAF;EACE,QAAQ,KAAR",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "test/fixtures/paths/in-paths.styl",
    "test/fixtures/basic.styl",
    "test/fixtures/source-map/index.styl",
  ],
  "sourcesContent": [
    ".other
	font-family serif",
    "border-radius()
  -webkit-border-radius arguments
  -moz-border-radius arguments
  border-radius arguments

body
  font 12px Helvetica, Arial, sans-serif

a.button
  border-radius 5px",
    "@import 'in-paths'
@import '../basic'

.some-class
  margin: 10px
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" options should generate nested source maps when value is "true": warnings 1`] = `[]`;

exports[`"sourceMap" options should generate source maps when the "devtool" value is "source-map": css 1`] = `
".other {
  font-family: serif;
}
body {
  font: 12px Helvetica, Arial, sans-serif;
}
a.button {
  -webkit-border-radius: 5px;
  -moz-border-radius: 5px;
  border-radius: 5px;
}
.some-class {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should generate source maps when the "devtool" value is "source-map": errors 1`] = `[]`;

exports[`"sourceMap" options should generate source maps when the "devtool" value is "source-map": source map 1`] = `
{
  "mappings": "AAAA;EACC,aAAY,MAAZ;;ACID;EACE,MAAmB,kCAAnB;;AAEF;EAPE,uBAAsB,IAAtB;EACA,oBAAmB,IAAnB;EACA,eAAc,IAAd;;ACAF;EACE,QAAQ,KAAR",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "test/fixtures/paths/in-paths.styl",
    "test/fixtures/basic.styl",
    "test/fixtures/source-map.styl",
  ],
  "sourcesContent": [
    ".other
	font-family serif",
    "border-radius()
  -webkit-border-radius arguments
  -moz-border-radius arguments
  border-radius arguments

body
  font 12px Helvetica, Arial, sans-serif

a.button
  border-radius 5px",
    "@import 'in-paths'
@import 'basic'

.some-class
  margin: 10px
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" options should generate source maps when the "devtool" value is "source-map": warnings 1`] = `[]`;

exports[`"sourceMap" options should generate source maps when value is "true" and the "devtool" value is "false": css 1`] = `
".other {
  font-family: serif;
}
body {
  font: 12px Helvetica, Arial, sans-serif;
}
a.button {
  -webkit-border-radius: 5px;
  -moz-border-radius: 5px;
  border-radius: 5px;
}
.some-class {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should generate source maps when value is "true" and the "devtool" value is "false": errors 1`] = `[]`;

exports[`"sourceMap" options should generate source maps when value is "true" and the "devtool" value is "false": source map 1`] = `
{
  "mappings": "AAAA;EACC,aAAY,MAAZ;;ACID;EACE,MAAmB,kCAAnB;;AAEF;EAPE,uBAAsB,IAAtB;EACA,oBAAmB,IAAnB;EACA,eAAc,IAAd;;ACAF;EACE,QAAQ,KAAR",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "test/fixtures/paths/in-paths.styl",
    "test/fixtures/basic.styl",
    "test/fixtures/source-map.styl",
  ],
  "sourcesContent": [
    ".other
	font-family serif",
    "border-radius()
  -webkit-border-radius arguments
  -moz-border-radius arguments
  border-radius arguments

body
  font 12px Helvetica, Arial, sans-serif

a.button
  border-radius 5px",
    "@import 'in-paths'
@import 'basic'

.some-class
  margin: 10px
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" options should generate source maps when value is "true" and the "devtool" value is "false": warnings 1`] = `[]`;

exports[`"sourceMap" options should generate source maps when value is "true": css 1`] = `
".other {
  font-family: serif;
}
body {
  font: 12px Helvetica, Arial, sans-serif;
}
a.button {
  -webkit-border-radius: 5px;
  -moz-border-radius: 5px;
  border-radius: 5px;
}
.some-class {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should generate source maps when value is "true": errors 1`] = `[]`;

exports[`"sourceMap" options should generate source maps when value is "true": source map 1`] = `
{
  "mappings": "AAAA;EACC,aAAY,MAAZ;;ACID;EACE,MAAmB,kCAAnB;;AAEF;EAPE,uBAAsB,IAAtB;EACA,oBAAmB,IAAnB;EACA,eAAc,IAAd;;ACAF;EACE,QAAQ,KAAR",
  "names": [],
  "sourceRoot": "",
  "sources": [
    "test/fixtures/paths/in-paths.styl",
    "test/fixtures/basic.styl",
    "test/fixtures/source-map.styl",
  ],
  "sourcesContent": [
    ".other
	font-family serif",
    "border-radius()
  -webkit-border-radius arguments
  -moz-border-radius arguments
  border-radius arguments

body
  font 12px Helvetica, Arial, sans-serif

a.button
  border-radius 5px",
    "@import 'in-paths'
@import 'basic'

.some-class
  margin: 10px
",
  ],
  "version": 3,
}
`;

exports[`"sourceMap" options should generate source maps when value is "true": warnings 1`] = `[]`;

exports[`"sourceMap" options should not generate source maps when the "devtool" value is "false": css 1`] = `
".other {
  font-family: serif;
}
body {
  font: 12px Helvetica, Arial, sans-serif;
}
a.button {
  -webkit-border-radius: 5px;
  -moz-border-radius: 5px;
  border-radius: 5px;
}
.some-class {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should not generate source maps when the "devtool" value is "false": errors 1`] = `[]`;

exports[`"sourceMap" options should not generate source maps when the "devtool" value is "false": warnings 1`] = `[]`;

exports[`"sourceMap" options should not generate source maps when value is "false" and the "devtool" value is "source-map": css 1`] = `
".other {
  font-family: serif;
}
body {
  font: 12px Helvetica, Arial, sans-serif;
}
a.button {
  -webkit-border-radius: 5px;
  -moz-border-radius: 5px;
  border-radius: 5px;
}
.some-class {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should not generate source maps when value is "false" and the "devtool" value is "source-map": errors 1`] = `[]`;

exports[`"sourceMap" options should not generate source maps when value is "false" and the "devtool" value is "source-map": warnings 1`] = `[]`;

exports[`"sourceMap" options should not generate source maps when value is "false": css 1`] = `
".other {
  font-family: serif;
}
body {
  font: 12px Helvetica, Arial, sans-serif;
}
a.button {
  -webkit-border-radius: 5px;
  -moz-border-radius: 5px;
  border-radius: 5px;
}
.some-class {
  margin: 10px;
}
"
`;

exports[`"sourceMap" options should not generate source maps when value is "false": errors 1`] = `[]`;

exports[`"sourceMap" options should not generate source maps when value is "false": warnings 1`] = `[]`;

exports[`"sourceMap" options should work and allow to override source maps options: css 1`] = `
"body {
  font: 12px Helvetica, Arial, sans-serif;
}
a.button {
  -webkit-border-radius: 5px;
  -moz-border-radius: 5px;
  border-radius: 5px;
}
/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImJhc2ljLnN0eWwiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBS0E7RUFDRSxNQUFtQixrQ0FBbkI7O0FBRUY7RUFQRSx1QkFBc0IsSUFBdEI7RUFDQSxvQkFBbUIsSUFBbkI7RUFDQSxlQUFjLElBQWQiLCJmaWxlIjoiYmFzaWMuY3NzIiwic291cmNlc0NvbnRlbnQiOlsiYm9yZGVyLXJhZGl1cygpXG4gIC13ZWJraXQtYm9yZGVyLXJhZGl1cyBhcmd1bWVudHNcbiAgLW1vei1ib3JkZXItcmFkaXVzIGFyZ3VtZW50c1xuICBib3JkZXItcmFkaXVzIGFyZ3VtZW50c1xuXG5ib2R5XG4gIGZvbnQgMTJweCBIZWx2ZXRpY2EsIEFyaWFsLCBzYW5zLXNlcmlmXG5cbmEuYnV0dG9uXG4gIGJvcmRlci1yYWRpdXMgNXB4Il19 */"
`;

exports[`"sourceMap" options should work and allow to override source maps options: errors 1`] = `[]`;

exports[`"sourceMap" options should work and allow to override source maps options: warnings 1`] = `[]`;
